### PR TITLE
erc4626

### PR DESCRIPTION
**Pull request by johhonn**
Created a generalized  bridge for erc4626 compatible vaults

### CODE REVIEW

File: src/bridges/erc4626/VaultBridge.sol [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        payable
+        override
+        returns (
+            uint256 outputValueA,
+            uint256 outputValueB,
+            bool isAsync
+        )
+    {
+        // // ### INITIALIZATION AND SANITY CHECKS
+        require(msg.sender == rollupProcessor, 'ExampleBridge: INVALID_CALLER');
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:64]: Since solidity 0.8.4 there is a support for custom errors. It would be nice if you used them instead of these require statements. Their advantage is that it decreases deployment costs because it doesn't store the error string in the bytecode but instead it only stores a 4 byte identifier pointing to the error definition in the ABI.

File: src/bridges/erc4626/VaultBridge.sol [hunk 2]
```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright 2020 Spilsbury Holdings Ltd
+pragma solidity >=0.6.10 <=0.8.10;
+pragma experimental ABIEncoderV2;
+
+import {SafeMath} from '@openzeppelin/contracts/utils/math/SafeMath.sol';
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:6]: All our bridges are targeting solidity >=0.8.4 so please set this version even for this bridge and remove this dependenct.

File: src/bridges/erc4626/VaultBridge.sol [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+   
+     */
+    // enter by deposit exact assets
+    function enter(
+        address vault,
+        address token,
+        uint256 amount
+    ) internal returns (uint256) {
+        IERC20 depositToken = IERC20(token);
+        depositToken.approve(vault, amount);
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:107]: It would be useful to pre-approve these tokens to safe gas. The bridge should never hold any ERC20s after the interaction settles so it is not a security risk. Please check our [liquity bridge](https://github.com/AztecProtocol/aztec-connect-bridges/blob/e53c3f2fe746aa02f680571f29c08f02a90d932d/src/bridges/liquity/StakingBridge.sol#L70) for inspiration.

File: src/bridges/erc4626/VaultBridge.sol [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        uint256 prev = IERC4626(vault).balanceOf(address(this));
+        IERC4626(vault).deposit(amount, address(this));
+        uint256 _after = IERC4626(vault).balanceOf(address(this));
+        IERC4626(vault).approve(rollupProcessor, _after.sub(prev));
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:111]: Pre-approve this as well.

File: src/bridges/erc4626/VaultBridge.sol [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    function exit(
+        address vault,
+        address token,
+        uint256 amount
+    ) internal returns (uint256) {
+        IERC20 withdrawToken = IERC20(token);
+        uint256 prev = withdrawToken.balanceOf(address(this));
+        IERC4626(vault).redeem(amount, msg.sender, address(this));
+        uint256 _after = withdrawToken.balanceOf(address(this));
+        withdrawToken.approve(rollupProcessor, _after.sub(prev));
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:133]: Pre-approve this as well.

File: src/bridges/multitoken/ExampleBridge.sol [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+pragma experimental ABIEncoderV2;
+
+import {SafeMath} from '@openzeppelin/contracts/utils/math/SafeMath.sol';
+
+import {IDefiBridge} from '../../interfaces/IDefiBridge.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import {AztecTypes} from '../../aztec/AztecTypes.sol';
+
+contract MutliBridgeContract is IDefiBridge {
```
Comment (benesjan) [inline src/bridges/multitoken/ExampleBridge.sol:13]: What is this contract here for? It looks just like some forgotten test file. If that's the case, please remove it.

File: src/test/erc4626/ERC4626.t.sol [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        console.log(rollupMapleShares, 'ouput amount');
+        assertEq(rollupMapleShares, 4996);
+        rollupProcessor.convert(address(vaultbridge), outputAsset, empty, inputAsset, empty, rollupMapleShares, 1, 0);
+        uint256 rollupMapleToken = maple.balanceOf(address(rollupProcessor));
+        console.log(rollupMapleToken, 'withdraw amount');
+        assertEq(rollupMapleToken, 4999);
+        //assertEq(depositAmount, rollupBeefy, 'Balances must match');
+    }
+
+    function assertNotEq(address a, address b) internal {
```
Comment (benesjan) [inline src/test/erc4626/ERC4626.t.sol:78]: This method is not used, so please remove it.

File: src/test/erc4626/ERC4626.t.sol [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import {DefiBridgeProxy} from './../../aztec/DefiBridgeProxy.sol';
+import {RollupProcessor} from './../../aztec/RollupProcessor.sol';
+import {IERC4626} from './../../bridges/erc4626/Interfaces/IERC4626.sol';
+import {VaultBridge} from './../../bridges/erc4626/VaultBridge.sol';
+import {AztecTypes} from './../../aztec/AztecTypes.sol';
+import {console} from '../console.sol';
+import '../../../lib/ds-test/src/test.sol';
+
+//Tested at block 14886873 may not work at other blocks
+contract ERC4626 is DSTest {
```
Comment (benesjan) [inline src/test/erc4626/ERC4626.t.sol:15]: Please extend Test class from forge-std library instead of importing DSTest directly. What is an advantage of using the class from forge-std is that there is a deal functionality which can be used for token minting. The advantage of deal(...) over the _setTokenBalance(...) is that it automatically detects the slot of mapping with balances.

Please check out [this example](https://github.com/AztecProtocol/aztec-connect-bridges/blob/e53c3f2fe746aa02f680571f29c08f02a90d932d/src/test/example/Example.t.sol#L14).

File: src/test/erc4626/ERC4626.t.sol [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    function assertNotEq(address a, address b) internal {
+        if (a == b) {
+            emit log('Error: a != b not satisfied [address]');
+            emit log_named_address('  Expected', b);
+            emit log_named_address('    Actual', a);
+            fail();
+        }
+    }
+
+    function _setTokenBalance(
```
Comment (benesjan) [inline src/test/erc4626/ERC4626.t.sol:87]: Replace this method with deal(...) from forge-std as mentioned in the comment above.

File: src/test/erc4626/ERC4626.t.sol [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    function setUp() public {
+        _aztecPreSetup();
+
+        vaultbridge = new VaultBridge(address(rollupProcessor));
+        //vaultbridge.validateVault(address(vault));
+        rollupProcessor.setBridgeGasLimit(address(vaultbridge), 1000000);
+    }
+
+    function testVaultBridge() public {
```
Comment (benesjan) [inline src/test/erc4626/ERC4626.t.sol:40]: Since the bridge is supposed to work for generic erc4626 vaults it would be useful to test it on more vaults on mainnet. This shouldn't take too much work as you would simply run this test over an array of vault addresses and the corresponding tokens. Finding more vaults should be fairly straightforward using etherscan's [smart contract search](https://etherscan.io/searchcontractlist?a=all&q=erc4626).

**Review by benesjan — CHANGES_REQUESTED** [review]
The overall code looks good. The changes I requested are mostly about making the code more similar to our other bridges.

Also please rebase this branch onto master and run yarn lint and try to fix the linting errors. Don't forget to fetch the dependencies again. Thank you.

File: src/bridges/erc4626/VaultBridge.sol (hunk 3 above)
Comment (johhonn) [reply src/bridges/erc4626/VaultBridge.sol:107]: That makes sense. The bridge I wrote i designed to work for any vault/ token pair.It is obviously impossible to pre approve them all. I could make a function that allows pre approvals for specific pairs. Would that make sense?

File: src/test/erc4626/ERC4626.t.sol (hunk 10 above)
Comment (johhonn) [reply src/test/erc4626/ERC4626.t.sol:40]: How many do you think would be adequate?

Comment (benesjan) [reply src/test/erc4626/ERC4626.t.sol]: I think 5 should be fine.

Comment (LHerskind) [reply src/test/erc4626/ERC4626.t.sol]: Please use vaults that have different input assets (ideally with different number of decimals), e.g., USDC, Dai, Eth, WBTC.

File: src/bridges/erc4626/VaultBridge.sol [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * @notice Internal Function used to unstake
+     * @dev This method redeems an exact number of shares into an erc4626 vault then approves the rollup processor
+     *
+     * @param vault address of erc4626 vault
+     * @param token address of the vault
+     * @param amount amount of shares to be redeemed
+     *
```
Comment (LHerskind) [inline src/bridges/erc4626/VaultBridge.sol:123]: Missing natspec for the return value

File: src/bridges/erc4626/VaultBridge.sol [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        try IERC4626(a).asset() returns (IERC20 token) {
+            return address(token) == b;
+        } catch {
+            return false;
+        }
+    }
+
+    /**
+     * @notice Internal Function used to stake
+     * @dev This method deposits an exact amount of asset into an erc4626 vault then approves the rollup processor
```
Comment (LHerskind) [inline src/bridges/erc4626/VaultBridge.sol:94]: Update the comment as you no longer approve

File: src/bridges/erc4626/VaultBridge.sol [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        IERC20 withdrawToken = IERC20(token);
+        uint256 prev = withdrawToken.balanceOf(address(this));
+        IERC4626(vault).redeem(amount, msg.sender, address(this));
+        uint256 _after = withdrawToken.balanceOf(address(this));
+        //withdrawToken.approve(rollupProcessor, _after.sub(prev));
+        return _after.sub(prev);
+    }
+
+    function approvePair(address vault, address token) public {
+        IERC20(token).approve(vault, type(uint256).max);
```
Comment (LHerskind) [inline src/bridges/erc4626/VaultBridge.sol:139]: Use `safeApprove` instead. Otherwise, certains assets (as USDT) might not be used because it don't follow the `approve` interface from the ERC20 standard.

File: src/test/erc4626/ERC4626.t.sol [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    function setUp() public {
+        _aztecPreSetup();
+
+        vaultbridge = new VaultBridge(address(rollupProcessor));
+        //vaultbridge.validateVault(address(vault));
+        rollupProcessor.setBridgeGasLimit(address(vaultbridge), 1000000);
+    }
+
+    function testVaultBridge1() public {
```
Comment (LHerskind) [inline src/test/erc4626/ERC4626.t.sol:41]: Please use property based testing (fuzzing) as part of the tests. Example here would be to have `depositAmount` as input, and then make it bound to be within some range.
```solidity
function testVaultBridge1(uint96 _depositAmount) public {
  uint256 depositAmount = bound(_depositAmount, 1000, type(uint96).max);
```

File: foundry.toml [hunk 15]
```diff
@@ -2,7 +2,7 @@
 src = 'src'
 out = 'out'
 libs = ['lib']
-remappings = ['ds-test/=lib/ds-test/src/', '@openzeppelin/=node_modules/@openzeppelin/']
+remappings = ['ds-test/=lib/ds-test/src/', '@openzeppelin/=node_modules/@openzeppelin/', 'forge-std/=lib/forge-std/src']
```
Comment (LHerskind) [inline foundry.toml:5]: Why is this changed?

File: src/test/erc4626/ERC4626.t.sol [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+            erc20Address: address(maple),
+            assetType: AztecTypes.AztecAssetType.ERC20
+        });
+        AztecTypes.AztecAsset memory outputAsset = AztecTypes.AztecAsset({
+            id: 1,
+            erc20Address: address(vault),
+            assetType: AztecTypes.AztecAssetType.ERC20
+        });
+
+        (uint256 outputValueA, uint256 outputValueB, bool isAsync) = rollupProcessor.convert(
```
Comment (LHerskind) [inline src/test/erc4626/ERC4626.t.sol:96]: The return values should be used in assertions.

File: src/test/erc4626/ERC4626.t.sol [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
```
Comment (LHerskind) [inline src/test/erc4626/ERC4626.t.sol:58]: The return values should be used in assertions.

**Comment by LHerskind** [comment]
Hi @johhonn, I took a deeper look and have some comments added below:
- Update to build on the latest master,
	- We have added some folders for general interfaces that can be used across bridges etc. The IERC4626 can be moved into that.
	- Also multiple extra solhint rules that would be great if you could enforce on your bridge
- Please inherit from the `BridgeBase` contract that was added. It helps on aligning the bridges with respect to error codes and authorizations etc
- Use `ErrorLib` errors when possible (added alon)
- Why does the `VaultBridge` inherit from `Ownable` I don't see you using it anywhere.
- The SafeMath is unused and can be removed. Also unneeded after 0.8.0.
- The `approvePair` will run into troubles if you have a vault that takes `USDT` as the input asset.
- In `enter` you set a `depositToken`, but the value is never used. Can be removed. Also, no need to use a `shares` variable, you can just return in instantly.
- In `exit` you can also just return the value from `redeem` instantly. 
- Add as a comment in the contract natspec that the bridge is stateless (will not hold tokens etc) itself, so approving is not as bad because there is nothing to pull.
- The `testPair` name is not that telling, might be better with `isValidPair()` or similar that tells more about what is done. Also, is it required to be public, or will it only ever be called by the contract itself? Can be internal instead then.
- For `enter` there is a mix of natspec and comment above the function name.
- Following the additional linter rules:
	- `enter` should be `_enter`
	- `exit` should be `_exit`
	-  `_enter` and `_exit` should be moved down as they are internal (follow ordering of functions).

File: src/bridges/erc4626/VaultBridge.sol [hunk 18]
```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright 2020 Spilsbury Holdings Ltd
+pragma solidity ^0.8.4;
+pragma experimental ABIEncoderV2;
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:4]: ABIEncoderV2 is default and no longer experimental so please remove the line above.

File: src/bridges/erc4626/VaultBridge.sol (hunk 11 above)
Comment (benesjan) [reply src/bridges/erc4626/VaultBridge.sol:126]: @johhonn this was not yet addressed. Please address all our comments before requesting a review.

File: src/bridges/erc4626/VaultBridge.sol [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+
+    // @notice This function always reverts because this contract does not implement async flow.
+    function finalise(
+        AztecTypes.AztecAsset calldata inputAssetA,
+        AztecTypes.AztecAsset calldata inputAssetB,
+        AztecTypes.AztecAsset calldata outputAssetA,
+        AztecTypes.AztecAsset calldata outputAssetB,
+        uint256 interactionNonce,
+        uint64 auxData
+    )
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:140]: Since you are inheriting from BridgeBase please remove this function. https://github.com/AztecProtocol/aztec-connect-bridges/blob/5c32f02782d5649287c626fdd01afb7591068578/src/bridges/base/BridgeBase.sol#L54

File: src/bridges/erc4626/VaultBridge.sol [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+    /**
+     * @notice Internal Function used to stake
+     * @dev This method deposits an exact amount of asset into an erc4626 vault
+     *
+     * @param vault address of erc4626 vault     
+     * @param amount amount of an asset to be burned
+   
+     */
+    // enter by deposit exact assets
```
Comment (benesjan) [inline src/bridges/erc4626/VaultBridge.sol:115]: Missing natspec for the return value here as well.

File: src/test/erc4626/ERC4626.t.sol [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+
+    IERC20 constant maple = IERC20(0x33349B282065b0284d756F0577FB39c158F935e6);
+    IERC20 constant fei = IERC20(0xc7283b66Eb1EB5FB86327f08e1B5816b0720212B);
+    IERC4626 constant tribeVault = IERC4626(0x4f93Df7Bc0421C9401fD3099cCE4AfE7678B0c63);
+
+    IERC4626 constant vault = IERC4626(0x4937A209D4cDbD3ecD48857277cfd4dA4D82914c);
+
+    function _aztecPreSetup() internal {
+        defiBridgeProxy = new DefiBridgeProxy();
+        rollupProcessor = new RollupProcessor(address(defiBridgeProxy));
```
Comment (benesjan) [inline src/test/erc4626/ERC4626.t.sol:31]: Please refactor this test so that [example test structure](https://github.com/AztecProtocol/aztec-connect-bridges/tree/master/src/test/bridges/example) is followed (separation to E2E and unit test file). Using DefiBridgeProxy and RollupProcessor directly in test has been deprecated.

**Review by benesjan — CHANGES_REQUESTED** [review]
I added few more comments. Please address them all. Can you please copy the files from this PR to a new branch which has the same head as master and send the PR again? The diff of this PR contains changes from outside of this PR and it makes it hard to review. This is caused by the fact that the changes from master has been merged instead of rebased on (see [our readme](https://github.com/AztecProtocol/aztec-connect-bridges/blob/janb/readme-update/README.md#writing-a-bridge)). Also please make CI pass in the new PR. Thank you

**Comment by LHerskind** [comment]
Closed as you have opened a new pr, #170, for the same task.
